File: 78.subsets.py
# @leet start


class Solution:
    def subsets(self, nums: list[int]) -> list[list[int]]:
        n = len(nums)
        ans = []

        def backtrack(first: int, curr: list[int]):
            ans.append(curr[:])

            for i in range(first, n):
                curr.append(nums[i])
                backtrack(i + 1, curr)
                curr.pop()

        backtrack(0, [])

        return ans


# @leet end

File: 78.subsets.cpp
// @leet start
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        // []
        // [], add 1 to each
        // [] [1], add 2 to each
        // etc
        vector<vector<int>> ss;
        int n = nums.size();

        ss.push_back({});
        for (int i = 0; i < n; ++i) {
            // copy ss into s2
            vector<vector<int>> s2 = ss;

            for (auto& s : s2) {
                s.push_back(nums[i]);
                ss.insert(ss.end(), s);
            }
        }

        return ss;
        
    }
    
};

// @leet end
